### ISSUE TITLE
[Bug Found] Logic Bugs (Nagashi, Haitei Kan, Ron)

### ISSUE THREAD

**Issue by hnsqdtt** [smly/RiichiEnv#107]
In the evenings, I have been developing a GPU-accelerated RL environment based on RiichiEnv for regression processing (built on Triton). During regression testing, I discovered the following correctness issues(Immediate regression testing has already revealed these bugs. I will continue updating this issue as further findings emerge, if needed.):
## 1. Nagashi Mangan Judgment
RiichiEnv only checks the `nagashi_eligible` flag, without verifying whether the discarded tiles were called (i.e., claimed by other players).

## 2. Riichi Path Ankan Gating Issue
In the Riichi path, RiichiEnv does not gate concealed kans (`ankan`) with `wall.len() > 14`, which allows ankan on the haitei draw.
According to standard rules, no kan is allowed on the last tile (haitei), because a kan requires drawing from the rinshan (dead wall), and the wall is already empty.

## 3. Incorrect Tile Used in Ron Settlement
RiichiEnv retrieves the `tile_id` for ron settlement from `self.last_discard`, but `last_discard` may be overwritten during the response handling process.
In the given seed:
- `last_discard` points to 9p (`tile_id=70`)
- The actual ron tile is 4p (`tile_id=51`)
As a result, `AgariCalculator` evaluates the hand using the wrong tile, returning:
```
han=0
yaku=[]
```
  Although I used an older, trimmed and modified core version of RiichiEnv for regression testing, I am confident these issues exist since the core logic has not been changed. However, I am unsure whether they have already been fixed in newer versions. Therefore, I am opening this issue here.
  Additionally, if you are interested in my GPU-accelerated environment, feel free to contact me. I currently do not have the bandwidth to adapt it to the latest version of RiichiEnv, and it only supports Tenhou-rule four-player red-dora mahjong.

---

私は夜間に、RiichiEnv をベースとした回帰処理用の GPU 加速 RL 環境（Triton ベース）を開発しています。回帰テスト中に、以下の正確性に関する問題を発見しました。(（即時の回帰テストにより、これらのバグはすでに確認されています。今後も追加の発見があれば、本 Issue を更新していきます。）)
### 1. 流し満貫の判定
RiichiEnv は `nagashi_eligible` フラグのみを確認しており、捨て牌が鳴かれたかどうか（他家に取られたかどうか）を検証していません。

### 2. 立直経路における暗槓のゲート不足
RiichiEnv の立直経路では、暗槓をチェックする際に `wall.len() > 14` の条件がなく、海底摸牌時に暗槓が可能になっています。
標準ルールでは海底牌での槓は不可です（槓後には嶺上牌を引く必要がありますが、牌山がすでに空のため不可能です）。

### 3. ロン精算時の牌参照ミス
RiichiEnv はロン精算時に `self.last_discard` から `tile_id` を取得していますが、`last_discard` は response 処理の過程で上書きされる可能性があります。
この seed では： - `last_discard` は 9p（`tile_id=70`）を指しています - 実際のロン牌は 4p（`tile_id=51`）ですその結果、`AgariCalculator` が誤った牌で計算を行い、

    han=0
    yaku=[]

が返されています。

回帰テストでは古いバージョンの簡略・改造済み RiichiEnv コアを使用していますが、コアロジック自体は変更していないため、これらの問題は存在すると確信しています。ただし、バージョン更新に伴い修正済みかどうかは確認できていません。そのため、本 Issue を提出しました。

また、私の GPU 加速環境にご興味があれば、ぜひご連絡ください。現在は最新バージョンの RiichiEnv に適合させる余裕がなく、天鳳ルール四人赤あり麻雀バージョンのみ対応しています。

**Comment by hnsqdtt** [comment]
Yep — as you said, a purely branching / state-machine style simulator usually doesn’t benefit much from GPU throughput. That said, I want to add a few points from my side.

First, it’s worth emphasizing that RiichiEnv basically doesn’t use precomputed lookup tables (a lot of these checks/counters could be sped up a lot via table lookups), which feels like a very real opportunity to cut overhead. Second, on a 32-core CPU, my training setup can generate about ~26k steps/s at batch=128 and it still scales roughly linearly with batch, while the GPU side doesn’t seem bottlenecked yet (I’d guess it only starts to saturate around batch 256–384; the forward-pass time vs. env time is roughly ~8:1). This suggests that as you scale up, the simulator’s absolute cost isn’t necessarily negligible (though it’s also possible my implementation/optimization is just bad — I’d love to align profiling methodology with what you see).

Another key motivation is exactly what you mentioned: keeping rollout / feature construction on-device to eliminate CPU↔GPU transfer and synchronization overhead.

Finally — ignoring the fact that I’m not as familiar with JAX 😄 — Triton feels attractive mainly because it gives finer-grained and more flexible control over key kernels, rather than forcing everything through a fully precompiled graph the way JAX often does. In your PPO runs, what does the wall-clock split between env and NN typically look like? Have you compared different batch sizes?

**Comment by hnsqdtt** [comment]
>  The ron tile is always correct.

About the issue “Incorrect Tile Used in Ron Settlement”:

the description is: “In `py.rs`, ron settlement prioritizes taking `last_discard`, but during robbed kan (抢杠) this value is the old discard.”
The proposed fix is: “In `py.rs`, swap the priority: first take `pending_kan.tile`, then fall back to `last_discard`.”

Please keep an eye on whether the new version implements it this way. If it has already been fixed, then it’s fine.

**Comment by smly** [comment]
I confirmed that the reported scenarios already work correctly in v0.4.8 and added regression tests in #226, including the chankan and daiminkan cases from your follow-up comments. Closing this issue once #226 is merged. Thanks again!
